pkg/operator/controllers/monitoring/monitoring_controller.go

```diff
@@ -0,0 +1,212 @@
+package monitoring
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+
+	"github.com/ghodss/yaml"
+	"github.com/sirupsen/logrus"
+	"github.com/ugorji/go/codec"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/retry"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	"github.com/Azure/ARO-RP/pkg/operator/controllers"
+)
+
+var monitoringName = types.NamespacedName{Name: "cluster-monitoring-config", Namespace: "openshift-monitoring"}
+
+// Config represents cluster monitoring stack configuration.
+// Reconciler reconciles retention and storage settings,
+// MissingFields are used to preserve settings configured by user.
+type Config struct {
+	api.MissingFields
+	PrometheusK8s struct {
+		api.MissingFields
+		Retention           string `json:"retention,omitempty"`
+		VolumeClaimTemplate struct {
+			api.MissingFields
+			Spec struct {
+				api.MissingFields
+				Resources struct {
+					api.MissingFields
+					Requests struct {
+						api.MissingFields
+						Storage string `json:"storage,omitempty"`
+					} `json:"requests,omitempty"`
+				} `json:"resources,omitempty"`
+			} `json:"spec,omitempty"`
+		} `json:"volumeClaimTemplate,omitempty"`
+	} `json:"prometheusK8s,omitempty"`
+}
+
+var defaultConfig = `prometheusK8s:
+  retention: 15d
+  volumeClaimTemplate:
+    spec:
+      resources:
+        requests:
+          storage: 100Gi
+`
+
+type Reconciler struct {
+	kubernetescli kubernetes.Interface
+	log           *logrus.Entry
+}
+
+func NewReconciler(log *logrus.Entry, kubernetescli kubernetes.Interface) *Reconciler {
+	return &Reconciler{
+		kubernetescli: kubernetescli,
+		log:           log,
+	}
+}
+
+func (r *Reconciler) Reconcile(request ctrl.Request) (ctrl.Result, error) {
+	// TODO(mj): controller-runtime master fixes the need for this (https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/reconcile/reconcile.go#L93) but it's not yet released.
+	ctx := context.Background()
+	if request.NamespacedName != monitoringName &&
+		request.Name != arov1alpha1.SingletonClusterName {
+		return reconcile.Result{}, nil
+	}
+
+	return reconcile.Result{}, retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		cm, isCreate, err := r.monitoringConfigMap(ctx)
+		if err != nil {
+			return err
+		}
+		if cm.Data == nil {
+			cm.Data = map[string]string{
+				"config.yaml": "",
```

Comment:
nit: this line is unnecessary.  Line 97 already has to handle the case where the "config.yaml" key doesn't exist and the zero value ("") is read, so you might as well remove it.